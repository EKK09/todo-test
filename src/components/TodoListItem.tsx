import React, { ChangeEvent } from 'react';
import { useTranslation } from 'react-i18next';
import useTodoListStore, { Todo } from 'src/store/todoList';
import styled from 'styled-components';

export interface TodoListItemProps {
  todo: Todo
}

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  border: 1px solid palevioletred;
  border-radius: 8px;
  margin-bottom: 4px;
  padding: 0 4px;
`;
const Title = styled.h3`
  font-size: 1.5rem;
  color: #727f8b;
`;
const TagsWrapper = styled.div`
  flex-grow: 1;
  display: flex;
  align-items: center;
  justify-content: center;
`;
const Tag = styled.span`
  background-color: CornflowerBlue;
  border-radius: 4px;
  color: white;
  padding: 4px 6px;
  margin: 4px;
`;
const CheckLabel = styled.span`
  color: #727f8b;
  font-weight: bolder;
`;

const TodoListItem = (props: TodoListItemProps): React.ReactElement => {
  const { todo } = props;
  const { t } = useTranslation();
  const updateTodo = useTodoListStore((state) => state.updateTodo);
  const handleCompletedChange = (event: ChangeEvent<HTMLInputElement>) => {
    updateTodo({
      ...todo,
      completed: event.target.checked,
    });
  };

  return (
    <Wrapper>
      <Title>
        {todo.title}
      </Title>
      <TagsWrapper>
        {todo.tags.map((tag) => <Tag key={tag}>{tag}</Tag>)}
      </TagsWrapper>
      <div>
        <CheckLabel>
          {t('todo.completed')}
          <input
            type="checkbox"
            checked={todo.completed}
            onChange={handleCompletedChange}
            data-testid="completed-checkbox"
          />
        </CheckLabel>
      </div>
    </Wrapper>
  );
};

export default TodoListItem;
