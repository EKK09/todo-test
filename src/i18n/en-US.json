{
    "global": {
        "appName": "TODO LIST",
        "logout": "Logout"
    },
    "loginPage": {
        "account": "Account",
        "password": "Password",
        "login": "Login"
    },
    "todo": {
        "completed": "Completed"
    }
}