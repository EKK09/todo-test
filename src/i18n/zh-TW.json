{
    "global": {
        "appName": "代辦事項"
    },
    "loginPage": {
        "account": "帳號",
        "password": "密碼",
        "login": "登入"
    },
    "todo": {
        "completed": "完成"
    }
}