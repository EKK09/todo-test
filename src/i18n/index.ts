import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import enUSResource from './en-US.json';
import zhTWResource from './zh-TW.json';

export enum Lang {
    EN_US = 'en-US',
    ZH_TW = 'zh-TW',
}

export const LANG_OPTION = {
  [Lang.EN_US]: 'English',
  [Lang.ZH_TW]: '中文',
};

const resources = {
  [Lang.EN_US]: {
    translation: enUSResource,
  },
  [Lang.ZH_TW]: {
    translation: zhTWResource,
  },
};

const getDefaultLang = (locale: string): Lang => {
  switch (locale) {
    case Lang.EN_US:
      return Lang.EN_US;

    case Lang.ZH_TW:
      return Lang.ZH_TW;

    default:
      return Lang.EN_US;
  }
};

const locale = window.navigator.language;

i18n
  .use(initReactI18next)
  .init({
    resources,
    lng: getDefaultLang(locale),
  });

export default i18n;
