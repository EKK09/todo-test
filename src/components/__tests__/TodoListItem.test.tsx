import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next';
import i18n, { Lang } from 'src/i18n';
import useTodoListStore, { Todo } from 'src/store/todoList';
import TodoListItem from '../TodoListItem';

describe('TodoListItem component', () => {
  const fooTodo: Todo = {
    id: 1234,
    title: 'fooTitle',
    tags: ['fooTag'],
    completed: false,
  };

  it('Render TodoListItem english version', async () => {
    i18n.changeLanguage(Lang.EN_US);
    const component = render(
      <I18nextProvider i18n={i18n}>
        <TodoListItem todo={fooTodo} />
      </I18nextProvider>,
    );

    expect(component.container).toMatchSnapshot();
  });

  it('Render TodoListItem chinese version', async () => {
    i18n.changeLanguage(Lang.ZH_TW);
    const component = render(
      <I18nextProvider i18n={i18n}>
        <TodoListItem todo={fooTodo} />
      </I18nextProvider>,
    );

    expect(component.container).toMatchSnapshot();
  });

  describe('TodoListItem checkbox status', () => {
    it('todo completed', async () => {
      const { getByTestId } = render(
        <I18nextProvider i18n={i18n}>
          <TodoListItem todo={{ ...fooTodo, completed: true }} />
        </I18nextProvider>,
      );
      const checkbox = getByTestId('completed-checkbox') as HTMLInputElement;

      expect(checkbox.checked).toBeTruthy();
    });

    it('todo not completed', async () => {
      const { getByTestId } = render(
        <I18nextProvider i18n={i18n}>
          <TodoListItem todo={{ ...fooTodo, completed: false }} />
        </I18nextProvider>,
      );
      const checkbox = getByTestId('completed-checkbox') as HTMLInputElement;

      expect(checkbox.checked).toBeFalsy();
    });
  });

  it('toggle todo completed status', async () => {
    const mockUpdateTodo = jest.fn();
    useTodoListStore.setState({ updateTodo: mockUpdateTodo });
    const { getByTestId } = render(
      <I18nextProvider i18n={i18n}>
        <TodoListItem todo={fooTodo} />
      </I18nextProvider>,
    );
    const checkbox = getByTestId('completed-checkbox');
    fireEvent.click(checkbox);

    expect(mockUpdateTodo).toBeCalledWith({
      ...fooTodo,
      completed: !fooTodo.completed,
    });
  });
});
