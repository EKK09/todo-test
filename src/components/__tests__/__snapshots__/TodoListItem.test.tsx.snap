// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem component Render TodoListItem chinese version 1`] = `
<div>
  <div
    class="sc-bdfBQB ilRLIg"
  >
    <h3
      class="sc-gsTEea jHzPNZ"
    >
      fooTitle
    </h3>
    <div
      class="sc-dlfnuX iIifOV"
    >
      <span
        class="sc-hKgJUU eywCOR"
      >
        fooTag
      </span>
    </div>
    <div>
      <span
        class="sc-eCstlR dhWdTc"
      >
        完成
        <input
          data-testid="completed-checkbox"
          type="checkbox"
        />
      </span>
    </div>
  </div>
</div>
`;

exports[`TodoListItem component Render TodoListItem english version 1`] = `
<div>
  <div
    class="sc-bdfBQB ilRLIg"
  >
    <h3
      class="sc-gsTEea jHzPNZ"
    >
      fooTitle
    </h3>
    <div
      class="sc-dlfnuX iIifOV"
    >
      <span
        class="sc-hKgJUU eywCOR"
      >
        fooTag
      </span>
    </div>
    <div>
      <span
        class="sc-eCstlR dhWdTc"
      >
        Completed
        <input
          data-testid="completed-checkbox"
          type="checkbox"
        />
      </span>
    </div>
  </div>
</div>
`;
